src/app/games/hive/HivePieceBehaviour.ts

```diff
@@ -0,0 +1,298 @@
+                'Spider': HivePieceBehaviourSpider.get(),
+                'SoldierAnt': HivePieceBehaviourSoldierAnt.get(),
+            });
+        }
+        return HivePieceBehaviour.INSTANCES.get()[piece.kind];
+    }
+
+    protected destinationIsEmpty(move: HiveMoveCoordToCoord, state: HiveState): boolean {
+        return state.getAt(move.end).isEmpty();
+    }
+
+    protected canSlide(state: HiveState, start: Coord, end: Coord): boolean {
+        // For the piece to slide from start to end,
+        // one of the two common neighbors of start and coord should be empty
+        const startNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(start));
+        const endNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(end));
+        const commonNeighbors: MGPSet<Coord> = startNeighbors.intersect(endNeighbors);
+        for (const neighbor of commonNeighbors) {
+            if (state.getAt(neighbor).isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public abstract moveValidity(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void>
+
+    public abstract getPossibleMoves(coord: Coord, state: HiveState): HiveMoveCoordToCoord[]
+
+}
+
+export class HivePieceBehaviourQueenBee extends HivePieceBehaviour {
+
+    private static INSTANCE: MGPOptional<HivePieceBehaviourQueenBee> = MGPOptional.empty();
+
+    public static get(): HivePieceBehaviourQueenBee {
+        if (this.INSTANCE.isAbsent()) {
+            this.INSTANCE = MGPOptional.of(new this());
+        }
+        return this.INSTANCE.get();
+    }
+
+
+    public moveValidity(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void> {
+        if (HexagonalUtils.areNeighbors(move.coord, move.end) === false) {
+            return MGPFallible.failure(HiveFailure.QUEEN_BEE_CAN_ONLY_MOVE_TO_DIRECT_NEIGHBORS());
+        }
+        if (this.destinationIsEmpty(move, state) === false) {
+            return MGPFallible.failure(HiveFailure.THIS_PIECE_CANNOT_CLIMB());
+        }
+        if (this.canSlide(state, move.coord, move.end) === false) {
+            return MGPFallible.failure(HiveFailure.MUST_BE_ABLE_TO_SLIDE());
+        }
+        return MGPFallible.success(undefined);
+    }
+
+    public getPossibleMoves(coord: Coord, state: HiveState): HiveMoveCoordToCoord[] {
+        const moves: HiveMoveCoordToCoord[] = [];
+        for (const neighbor of HexagonalUtils.neighbors(coord)) {
+            if (state.getAt(neighbor).isEmpty()) {
+                moves.push(HiveMoveCoordToCoord.from(coord, neighbor).get());
+            }
+        }
+        return moves;
+    }
+}
+
+export class HivePieceBehaviourBeetle extends HivePieceBehaviour {
+
+    private static INSTANCE: MGPOptional<HivePieceBehaviourBeetle> = MGPOptional.empty();
+
+    public static get(): HivePieceBehaviourBeetle {
+        if (this.INSTANCE.isAbsent()) {
+            this.INSTANCE = MGPOptional.of(new this());
```

Comment:
this(), berk :D "it's irregular!" comme dit le libraire dans GoT

Reply:
pourtant c'est plus régulier, ça permet de copier coller tous les .get entre comportement de chaque pièce !

Reply:
Non euuuh, c'pas copier colland içi :D c'est moooooche :D !

Reply:
j'avais oublié de dire que j'avais corrigé :D Mais sinon, c'est plus générique comme code ! Si tu renommes la classe pouf une erreur de moins à résoudre